### PR TITLE
fix(suite): fix number input not reflecting form changes properly

### LINKED ISSUE #5928: Format input numbers
- [x] ignore all non-numeric symbols except `,` and `.`
- [x] first used `,` or `.` interpret as a decimal separator
- [x] ignore other usage of `,` or `.`
- [x] format number according the language settings. For example:
    - english → `1,234.56` 
    - czech → `1 234,56`
    - spanish → `1.234,56`
- [x] https://github.com/trezor/trezor-suite/issues/5272
- [x] review validation messages

### LINKED ISSUE #7984: Send dialogue fiat field not auto filled
**Describe the bug**
Fiat field is never filled when you enter amount manually.
When you select send max it works as expected.
When you edit amount it works. 
But when you just enter amount and wait it never gets filled.

**Steps to reproduce:**
1. Navigate to `Send` dialogue
2. Fill in `Address` and `Amount` fields
3. Observe fiat field 

**Info:**
 - Suite version: desktop 23.4.0 (21653072a7c14c04156cfc6992cec5b582b00a99)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuiteDev/23.4.0 Chrome/110.0.5481.104 Electron/23.1.1 Safari/537.36
 - OS: MacIntel
 - Screen: 1440x900
 - Device: model T 2.5.3 Universal (revision 2f03ace311584988d5aeab58fd1acf24ef54711a)
 - Transport: BridgeTransport 2.0.32

**Screenshots:**
image

image

### PR DESCRIPTION
**Pull request by dahaca**
## Description

* the Fiat field (as well as any other place with a similar pattern of handlers) is properly reacting to the Amount field changes

It's an ugly solution, I was not sure there is another quick one. Basically, inputs hooked onto the `react-hook-form` should ideally not have complex `onChange` logic, and rather use the `watch` method to ensure that the form state is fully updated. If you log the `error` in `handleInputChange` in the current code, you will see that the value there is outdated. Between updating memoized callbacks and form state some time passes and `handleInputChange` is not called with a new `errors` state. Could be avoided by using `watch`. But since I do not know where else this bug occurs I went for a more universal but ugly solution.

If anyone have any ideas – they are definitely welcome!  @matejkriz @Nodonisko @komret 

Related: https://github.com/trezor/trezor-suite/issues/5928
Related: [#7984](https://github.com/trezor/trezor-suite/issues/7984)

### CODE REVIEW

**Comment by sime** [comment]
I spotted fiat values not appearing and this solves it.

My 21sats about the uglyiness, should we consider writing tests around this?

File: packages/suite/src/components/suite/NumberInput.tsx
Comment (komret) [inline packages/suite/src/components/suite/NumberInput.tsx:1]: The solution in d8a763f226d04620bfcff6142799dd502f12838f is not ideal, indeed. Are you going to try some other approach later, @Leo? I think this solution is only good as a hotfix.

I found a small bug when deleting digits. I get an error, but the fiat value still recalculates. It should just disappear when going from 600 to 60.

https://user-images.githubusercontent.com/42465546/230130331-fe238013-2dd5-4304-82b4-727b12eae02d.mov

**Review by komret — CHANGES_REQUESTED** [review]

File: packages/suite/src/components/suite/NumberInput.tsx
Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: https://github.com/trezor/trezor-suite/pull/7995/commits/562804db1085169d4df7ff5bb68f4363a4e30c54

**Review by komret — APPROVED** [review]
Approving as a temporary solution. We should come up with a proper fix for the next release. No `setTimeout` hacks - we'll probably have to adjust the `onChange` handler in `Amount`.

**Review by komret — CHANGES_REQUESTED** [review]
One more thing: when you cut only a part of the value, the whole value is removed.

File: packages/suite/src/components/suite/NumberInput.tsx
Comment (komret) [inline packages/suite/src/components/suite/NumberInput.tsx:1]: I found some issues regarding selection, see:

This one is not related to cutting, I missed it before. When you DEL a character, the cursor end up in a wrong position (should be after "2"):

https://user-images.githubusercontent.com/42465546/231141766-d926b992-1250-4c1f-bb16-a798e60b185f.mov

Cutting, cursor should be after "4":

https://user-images.githubusercontent.com/42465546/231141769-19ec64cf-f3ac-4889-a930-185d6f03fc79.mov

Cutting, cursor should be after "1":

https://user-images.githubusercontent.com/42465546/231141771-3ee1e96b-1fbf-4e13-9a9b-697d2b645bd7.mov

Also typos on line 193.

File: packages/suite/src/components/suite/NumberInput.tsx [hunk 1]
```diff
@@ -211,26 +240,47 @@ export const NumberInput = ({
...
     // copy the non-formatted value
     const handleCopy = useCallback(
         (e: React.ClipboardEvent<HTMLInputElement>) => {
-            e.clipboardData.setData('text/plain', cleanValueString(displayValue, locale));
+            if (!inputRef.current) {
+                return;
+            }
+
+            const { selectionStart, selectionEnd, value } = inputRef.current;
+            if (selectionStart === null || selectionEnd == null) {
```
Comment (komret) [inline packages/suite/src/components/suite/NumberInput.tsx]: `==" is here on purpose?

Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx]: https://github.com/trezor/trezor-suite/pull/7995/commits/148d58e0e8a065a294e2ca4894db4aeb0df76e69

File: packages/suite/src/components/suite/NumberInput.tsx
Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: https://github.com/trezor/trezor-suite/pull/7995/commits/148d58e0e8a065a294e2ca4894db4aeb0df76e69

File: packages/suite/src/components/suite/NumberInput.tsx [hunk 2]
```diff
@@ -337,6 +357,8 @@ export const NumberInput = ({
     const handleKeyDown = useCallback(
         (e: React.KeyboardEvent<HTMLInputElement>) => {
             const pressedKey = e.key;
+            setPressedKey(pressedKey);
+            console.log(pressedKey);
```
Comment (komret) [inline packages/suite/src/components/suite/NumberInput.tsx]: Forgotten log

File: packages/suite/src/components/suite/NumberInput.tsx [hunk 3]
```diff
@@ -273,12 +292,13 @@ export const NumberInput = ({
...
 
             const { selectionStart, selectionEnd, value } = inputRef.current;
-            if (selectionStart === null || selectionEnd == null) {
+            if (selectionStart === null || selectionEnd === null) {
                 return;
             }
 
             const resultString = value.substring(0, selectionStart) + value.substring(selectionEnd);
             handleChange(resultString);
+            handleChange(resultString);
```
Comment (komret) [inline packages/suite/src/components/suite/NumberInput.tsx:302]: Is this on purpose? See line above.

File: packages/suite/src/components/suite/NumberInput.tsx [hunk 4]
```diff
@@ -185,12 +186,14 @@ export const NumberInput = ({
 
             const { groupSeparatorCharCode } = getLocaleSeparatorCharCodes(locale);
             const previousDisplayValue = previousDisplayValueRef.current;
-
-            // handle deleting a thousands separator with a DEL key (should not be possible with a Backspace),
+            // CMD+D on Mac
```
Comment (komret) [inline packages/suite/src/components/suite/NumberInput.tsx]: Should be Ctrl+D

File: packages/suite/src/components/suite/NumberInput.tsx
Comment (komret) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: The first issue is resolved, but not the latter two. Note that in the examples above, I am cutting (CMD+X) the values, not deleting.

There is one more bug. When you delete the digit after separator (in the example below, 3), validation does not trigger. It still says I don't have enough funds, although I do have more than 1245 sat. It triggers correctly when I delete any other digit.
https://user-images.githubusercontent.com/42465546/231245926-5df9b572-89f3-46fb-95b0-e4609fbdaea4.mov

Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: ca226e56e

File: packages/suite/src/components/suite/NumberInput.tsx (hunk 2 above)
Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx]: ca226e56e

File: packages/suite/src/components/suite/NumberInput.tsx (hunk 3 above)
Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx:302]: ca226e56e

File: packages/suite/src/components/suite/NumberInput.tsx (hunk 4 above)
Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx]: ca226e56e

File: packages/suite/src/components/suite/NumberInput.tsx
Comment (komret) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: See the issue with validation in the above comment, that one is not resolved yet.

Comment (komret) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: There is also a small inconsistency when deleting characters - the cursor sometimes end up before the separator, sometimes after it. It should alway be before the separator.
https://user-images.githubusercontent.com/42465546/231393853-36c59106-fac3-4909-8a53-abec0f777f41.mov

Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: first comment - 8d7f7945f
second comment - f5b3cca9f

Comment (komret) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: Good job, the above issues are fixed.

I discovered one more: when you attempt to remove a thousands separator by selecting it and then deleting, it should not do anything and cursor should move before the separator. That's how it works with backspace. With DEL though, the digit behind the separator is deleted instead.

https://user-images.githubusercontent.com/42465546/231451359-771b0c48-367e-4325-b586-be66a2bcfb1f.mov

Comment (dahaca) [reply packages/suite/src/components/suite/NumberInput.tsx:1]: 0a5529249

**Review by komret — APPROVED** [review]
It seems to me that it works correctly now. Still, some space for follow-up in a future release:
- more tests to cover selection issues etc.
- enabling undo/redo
- removing the `setTimeout` hack (this will need some changes in the parent components)

**Pull request merged**